ERCS/erc-7527.md

```diff
@@ -0,0 +1,516 @@
+---
+eip: 7527
+title: Token Bound Function Oracle AMM Contract
+description: Interfaces that wrap/unwrap between fungible tokens and non-fungible tokens based on an embedded Function Oracle AMM
+author: Elaine Zhang (@lanyinzly) <lz8aj@virginia.edu>, Jerry <jerrymindflow@gmail.com>, Amandafanny <amandafanny200@gmail.com>, Shouhao Wong (@wangshouh) <wongshouhao@outlook.com>, 0xPoet <0xpoets@gmail.com>
+discussions-to: https://ethereum-magicians.org/t/eip-7527-token-bound-function-oracle-amm-contract/15950
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-09-03
+requires: 165, 721
+---
+
+
+## Abstract
+
+This proposal outlines interfaces for a Function Oracle capable of converting fungible tokens into a single non-fungible token each instance and then reverting the non-fungible token back to fungible tokens. The pre-configured Function Oracle comes equipped with its own Automated Market Maker (AMM), enabling a clearing mechanism for voucher issuers.
+
+## Motivation
+
+The motivation behind a wrapping interfaces with embedded Function Oracle AMM is to achieve decentralized endorsement and thus quantifiable decentralized credit with liquidity in nature. This creates an applicable infrastructure for credit creation, transmission, and oracle in payment, social, and financial context. Decentralized credit can be implemented in the form of Non-fungible tokens, compatible with current Ethereum environment, which also expands its application scenarios. The concept of decentralized credit is highly integrated with dApps development. It aims to foster a more trustless environment.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+Contract Interfaces: 
+
+Three interfaces are included here: Agency, App, and Factory. 
+
+Agency and App MAY be seperately implemented, but they SHALL be mutually bounded and not upgradable after initialization. Or, Agency and App MAY be both implemented in one single contract. 
```

Comment:
This paragraph is a bit under-specified. What does "mutually bounded" mean?

Reply:
`onlyAgency()` and  `onlyApp()` modifiers here intended to ensure "mutually bounded."

`App` SHALL implement `onlyAgency()` modifier and `Mint` and `Burn` SHALL apply `onlyAgency()` as a modifier, which restricts calls to `Mint` and `Burn` only have effect if they are called through the corresponding `Agency`.

`Agency` is OPTIONAL to implement `onlyApp()`.